# Supply Chain Analysis
This Jupyter Notebook is related to the Supply Chain Analysis dataset by Shashwat Tiwari.

- Dataset: https://www.kaggle.com/datasets/shashwatwork/dataco-smart-supply-chain-for-big-data-analysis

#### About the dataset

A DataSet of Supply Chains used by the company DataCo Global was used for the analysis. Dataset of Supply Chain , which allows the use of Machine Learning Algorithms and R Software.
Areas of important registered activities : Provisioning , Production , Sales , Commercial Distribution.It also allows the correlation of Structured Data with Unstructured Data for knowledge generation.

Type Data :
Structured Data : DataCoSupplyChainDataset.csv
Unstructured Data : tokenized_access_logs.csv (Clickstream)

Types of Products : Clothing , Sports , and Electronic Supplies

Additionally it is attached in another file called DescriptionDataCoSupplyChain.csv, the description of each of the variables of the DataCoSupplyChainDatasetc.csv.
Acknowledgements & Source

Fabian Constante,

Fabian Constante
Instituto Politecnico de Leiria Escola Superior de Tecnologia e Gestao
Contribution: Master interested in topics related to Big Data
Fernando Silva, António Pereira

https://data.mendeley.com/datasets/8gx2fvg2k6/5

## Loading the data

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns', 600)

df = pd.read_csv('DataCoSupplyChainDataset.csv',header= 0,encoding='unicode_escape')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [6]:
df.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country Customer Email Customer Fname  Customer Id Customer Lname  \
0      Puerto Rico      XXXXXXXXX          Cally        20755       Holloway   
1      Puerto Rico      XXXXXXXXX          Irene        19492           Luna   
2          EE. UU.      XXXXXXXXX        Gillian        19491      Maldonado   
3          EE. UU.      XXXXXXXXX           Tana        19490           Tate   
4      Puerto Rico      XXXXXXXXX           Orli        19489      Hendricks   

  Customer Password Customer Segment Customer State           Customer Street  \
0         XXXXXXXXX         Consumer             PR  5365 Noble Nectar Island   
1         XXXXXXXXX         Consumer             PR          2679 Rustic Loop   
2         XXXXXXXXX         Consumer             CA      8510 Round Bear Gate   
3         XXXXXXXXX      Home Office             CA           3200 Amber Bend   
4         XXXXXXXXX        Corporate             PR  8671 Iron Anchor Corners   

   Customer Zipcode  Department Id Department Name   Latitude   Longitude  \
0             725.0              2         Fitness  18.251453  -66.037056   
1             725.0              2         Fitness  18.279451  -66.037064   
2           95125.0              2         Fitness  37.292233 -121.881279   
3           90027.0              2         Fitness  34.125946 -118.291016   
4             725.0              2         Fitness  18.253769  -66.037048   

         Market  Order City Order Country  Order Customer Id  \
0  Pacific Asia      Bekasi     Indonesia              20755   
1  Pacific Asia     Bikaner         India              19492   
2  Pacific Asia     Bikaner         India              19491   
3  Pacific Asia  Townsville     Australia              19490   
4  Pacific Asia  Townsville     Australia              19489   

  order date (DateOrders)  Order Id  Order Item Cardprod Id  \
0         1/31/2018 22:56     77202                    1360   
1         1/13/2018 12:27     75939                    1360   
2         1/13/2018 12:06     75938                    1360   
3         1/13/2018 11:45     75937                    1360   
4         1/13/2018 11:24     75936                    1360   

   Order Item Discount  Order Item Discount Rate  Order Item Id  \
0            13.110000                      0.04         180517   
1            16.389999                      0.05         179254   
2            18.030001                      0.06         179253   
3            22.940001                      0.07         179252   
4            29.500000                      0.09         179251   

   Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
0                    327.75                     0.29                    1   
1                    327.75             

In [7]:
df.shape

(180519, 53)

In [8]:
df.columns

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

## Feature description



## Data preprocessing
Evaluating missing values and duplicates

### Missing values
The Order Zipcode and Product Description features have 155,679 and 180,519 missing values, respectively.

In [9]:
df.isna().sum()

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Email                        0
Customer Fname                        0
Customer Id                           0
Customer Lname                        8
Customer Password                     0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0


### Duplicated data
The sum of duplicated data for the dataset is zero.<br>
There are no duplicates in this dataset.

In [10]:
df.duplicated().sum()

0

In [11]:
df_basic_statistics = pd.DataFrame(df.describe()).transpose()

df_basic_statistics['Interval'] = df_basic_statistics['max'] - df_basic_statistics['min']

df_basic_statistics = df_basic_statistics[['count', 'mean','std','min','max','Interval','25%','50%','75%']]

print("Dataset basic statistics.")

df_basic_statistics

Dataset basic statistics.


count          mean           std  \
Days for shipping (real)       180519.0      3.497654      1.623722   
Days for shipment (scheduled)  180519.0      2.931847      1.374449   
Benefit per order              180519.0     21.974989    104.433526   
Sales per customer             180519.0    183.107609    120.043670   
Late_delivery_risk             180519.0      0.548291      0.497664   
Category Id                    180519.0     31.851451     15.640064   
Customer Id                    180519.0   6691.379495   4162.918106   
Customer Zipcode               180516.0  35921.126914  37542.461122   
Department Id                  180519.0      5.443460      1.629246   
Latitude                       180519.0     29.719955      9.813646   
Longitude                      180519.0    -84.915675     21.433241   
Order Customer Id              180519.0   6691.379495   4162.918106   
Order Id                       180519.0  36221.894903  21045.379569   
Order Item Cardprod Id         180519.0    692.509764    336.446807   
Order Item Discount            180519.0     20.664741     21.800901   
Order Item Discount Rate       180519.0      0.101668      0.070415   
Order Item Id                  180519.0  90260.000000  52111.490959   
Order Item Product Price       180519.0    141.232550    139.732492   
Order Item Profit Ratio        180519.0      0.120647      0.466796   
Order Item Quantity            180519.0      2.127638      1.453451   
Sales                          180519.0    203.772096    132.273077   
Order Item Total               180519.0    183.107609    120.043670   
Order Profit Per Order         180519.0     21.974989    104.433526   
Order Zipcode                   24840.0  55426.132327  31919.279101   
Product Card Id                180519.0    692.509764    336.446807   
Product Category Id            180519.0     31.851451     15.640064   
Product Description                 0.0           NaN           NaN   
Product Price                  180519.0    141.232550    139.732492   
Product Status                 180519.0      0.000000      0.000000   

                                       min            max       Interval  \
Days for shipping (real)          0.000000       6.000000       6.000000   
Days for shipment (scheduled)     0.000000       4.000000       4.000000   
Benefit per order             -4274.979980     911.799988    5186.779968   
Sales per customer                7.490000    1939.989990    1932.499990   
Late_delivery_risk                0.000000       1.000000       1.000000   
Category Id                       2.000000      76.000000      74.000000   
Customer Id                       1.000000   20757.000000   20756.000000   
Customer Zipcode                603.000000   99205.000000   98602.000000   
Department Id                     2.000000      12.000000      10.000000   
Latitude                        -33.937553      48.781933      82.719486   
Longitude                      -158.025986     115.263077     273.289062   
Order Customer Id                 1.000000   20757.000000   20756.000000   
Order Id                          1.000000   77204.000000   77203.000000   
Order Item Cardprod Id           19.000000    1363.000000    1344.000000   
Order Item Discount               0.000000     500.000000     500.000000   
Order Item Discount Rate          0.000000       0.250000       0.250000   
Order Item Id                     1.000000  180519.000000  180518.000000   
Order Item Product Price          9.990000    1999.989990    1989.999990   
Order Item Profit Ratio          -2.750000       0.500000       3.250000   
Order Item Quantity               1.000000       5.000000       4.000000   
Sales                             9.990000    1999.989990    1989.999990   
Order Item Total                  7.490000    1939.989990    1932.499990   
Order Profit Per Order        -4274.979980     911.799988    5186.779968   
Order Zipcode                  1040.000000   99301.000000   98261.000000   
Pro

In [24]:
df['order date (DateOrders)'] = pd.to_datetime(df['order date (DateOrders)'])
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'])

## Exploratory Data Analysis (EDA)


In [ ]:
object_type_features = df.select_dtypes(include='object').columns.tolist()

for i in object_type_features: #list of features of type object
    print(f"------------------------------\n{i}\n------------------------------\n")
    print(df[i].value_counts(),
    "\n------------------------------\n\n")

### Customer Insights
First, let's explore some information about the customers.

#### Number of Unique Customers

#### Most Frequent Buyers

#### Average and Total Sales per Customer
Average and total Sales per Customer, grouped by:

- Country, State, City

- Segment (Consumer, Corporate, Home Office)

- Sales distribution (histogram/boxplot).

#### Total Sales per Customer

### Shipping Performance
Compare scheduled vs actual shipping days:

- df["Shipping_Delay"] = df["Days for shipping (real)"] - df["Days for shipment (scheduled)"]

Analyze delays across:

- Markets (e.g., LATAM vs USCA)

- Shipping Modes

- Delivery Status

- Proportion of Late vs On-time deliveries by Market, Category, etc.

### Sales and Profits
Most/least profitable products, categories, and customers.

Average Order Item Profit Ratio by:

- Market

- Product Category

- Shipping Mode

Correlation heatmap for numerical variables (e.g., sales, quantity, profit, discount rate).

### Geospatial Analysis
Plot orders by latitude/longitude.

Choropleth (maybe?) maps for:

- Sales by country/state.

- Delivery performance per region.

### Time Series Insights
Monthly/weekly trends of:

- Orders

- Sales

- Late deliveries

Seasonality or patterns in shipping delays.

## Statistical & Predictive Modeling

### 1. Classification Models
Predict Late Delivery

Target: Late_delivery_risk

Features (examples):

- Days for shipment (scheduled)

- Shipping Mode

- Market

- Product Category

- Order Item Discount Rate

Models to try:

- Logistic Regression

- Random Forest

- XGBoost

- LightGBM ?

Classification metrics: accuracy, precision, recall, F1-score, ROC-AUC

### 2. Regression Models
Predict Benefit per order or Order Profit Per Order

Features:

- Sales

- Order Item Quantity

- Order Item Discount Rate

- Product Category

- Shipping Mode

Model to try:

- Linear Regression

- Lasso/Ridge Regression?

- Random Forest Regressor?

### 3. Clustering

Customer Segmentation using:

Features:

- Sales
- Profit
- Location
- Segment

Product Clustering:

- Price
- Sales volume
- Discount rate
- Profit ratio

Models to try:
- KMeans
- DBSCAN
- Hierarchical Clustering?
- Visualize in 2D using PCA or t-SNE.

### 4. Market Basket Analysis

- Use Apriori algorithm or FP-Growth to find products that are frequently bought together.

- Use the Order Id to group products.

### 5. Anomaly Detection?

Spot unusual:
- Shipping delays

- Discounts

- Order profit ratios

Try Isolation Forest or One-Class SVM?